### ISSUE TITLE
Streams don't fail on WebSocket idle timeout on the server-side

### ISSUE THREAD

**Issue by MichaelZinsmaier** [akka/akka-http#1026]
Hi all,

when upgrading from Akka HTTP 10.0.0 to 10.0.5 I noticed that in case of a server side idle timeout the WebSocket server implementation does not fail the server side streams anymore.

Before 10.0.1 the server side stream terminated with
  - downstream finished
 -  upstream failure Timeout Exception

now it seems to be
 -  downstream finished
  - downstream finished

best take a look at the reproducer
https://github.com/MichaelZinsmaier/akka-http-timeout-bug

I think the 10.0.0 behavior is the better solution since server side code might want to react to a idle timeout in a specific way and propagating it with onError makes sense.

Best Michael

I also found this ticket but not sure if these issues are related
https://github.com/akka/akka-http/issues/1012

**Comment by jrudolph** [comment]
Thanks for the report and the reproducer, @MichaelZinsmaier. 

If that worked before, it seems to have been the case accidentally as well:

```scala
  val greeterWebSocketService =
    Flow[Message]
      .via(new MyMonitorFlow("in"))
      .mapConcat {
        case tm: TextMessage.Strict => {
          val text = tm.text
          println(s"request from $text")
          TextMessage.Strict(s"Hello $text") :: Nil
        }
        case bm: BinaryMessage =>
          bm.dataStream.runWith(Sink.ignore)
          Nil
      }
      .via(new MyMonitorFlow("out"))
```

If you have seen a "out downstream finish", it will propagate through your own stream to the "in" side as well. It seems in the 10.0.0 case, this propagation was slower than the failure managed to be processed on the in side. This was a race condition and may still be one. (It's a pity that we cannot transport downstream errors btw)

If you need access to the error message you can try to prevent propagation of the cancellation for a while to get access to the error message. We had a similar problem in our infrastructure that cancellation would RST TCP connections to early. Recently we introduced a [`DelayCancellation` stage](https://github.com/akka/akka-http/blob/master/akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala#L166) that does implement exactly this functionality. Can you try copy this stage to your code and see you would profit from that as well?

**Comment by MichaelZinsmaier** [comment]
Hi jrudolph,

I added a DelayCancellationFlow to the project, but the issue is still there.
Plus I added a new test based on a BidiPipe IdleTimeout.scala.
(remember that trick (-? something you helped me with still incredible useful)

At least in the BidiPipe example the flows on server side are not connected,
the cancellation should not propagate. I guess it could still propagate on client side but I
added Delay Stages there.

Plus I tested a bit more with our local environment and captured two Wireshark dumps
(images in the repo). To me it looks as if the error propagation did change somehow.
With 10.0.0 a idle timeout leads to a timely RST by the server with 10.0.5 that's not the case.

Still I might be wrong. I would like to test a bit more but currently have no time. If you can take a look I would appreciate that. Else I try to revisit soonish.

Best Michael

**Comment by jrudolph** [comment]
I can reproduce the behavior. I think we should first try to specify what the expected behavior should be and then implement it.

AFAICS the culprit is here: https://github.com/akka/akka-http/blob/3c966af2255b58b02fbdacd6abec89c47f78ce48/akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala#L660-L660

What happens here is that the idle-timeout will trigger both cancel and failure, and the switch stage will just complete the WS pipeline if it received the cancellation first.

**Comment by MichaelZinsmaier** [comment]
Hi, glad you could reproduce
regarding correct behavior I think there are at least two different cases to distinguish

idle timeout used
- "raw TCP" like, timeout probably set to a high value, no activity over a longer time
- using WS heartbeat feature, timeout probably set to a low value, heartbeat is skipped => timeout

The 2nd case is our use case an here since the application uses heartbeats + timeouts to detect half open (dropped) connections raising an error seems to be the correct approach. For the first case I think completion or error might be ok since probably nobody actually cares and this is just cleaning up of unused resources.

thus I think error as a common behavior (the 10.0.0) behavior is preferable

**Comment by ochrons** [comment]
I may have observed similar problem when a WS stream is handled on the server by a `Sink.actorRef`. When this actor is terminated, the Websocket connection is left alive (from the client point of view). Client happily sends Ping messages to WS until its own watchdog notices that it's not getting any Pongs.

This is within the ScalaFiddle project.

**Comment by MichaelZinsmaier** [comment]
Update for Akka Http 10.0.8

I did run the tests again and there is some notable difference

- the server now sends a RST towards the client at the point where the timeout hits
  (I think that's actually really good since the connection is effectively dead at this point)

- however the connected streams of the server still don't get a Failure, instead upstream finishes

I would argue that internal propagation of the error would be better since it allows to react in case PingPong is used (see above) plus client and server behavior would be the same downstream finish, upstream failed

still no time to work on it myself though ...

Best Michael

**Comment by jrudolph** [comment]
Interesting, thanks for the update.

**Comment by jrudolph** [comment]
@ochrons thanks for chiming in. Is there public code somewhere where this can be observed?

**Comment by ochrons** [comment]
Yes, in https://github.com/scalafiddle/scalafiddle-core (router is the WS server, compiler-server is the WS client)

https://github.com/scalafiddle/scalafiddle-core/blob/master/compiler-server/src/main/scala/scalafiddle/compiler/CompileActor.scala#L49-L54 is the work around (checking connection is alive with Ping/Pong)

**Comment by jrudolph** [comment]
I retested this with 10.2.x and it now seems to work as intended. Most likely this was caused by the dreaded cancellation race that was finally fixed in #3022.

Let's treat this as a duplicate to the later #3247 which has been fixed by now.
